File: SOLID/LSP(LiskovSubstitutionPrinciple)/ReturnTypeRule.cpp
#include <bits/stdc++.h>
using namespace std;

class Animal
{
};
class Dog : public Animal
{
};

class Parent
{
public:
    virtual Animal *getAnimal()
    {
        cout << "Parent return animal instance" << endl;
        return new Animal();
    }
};

class Child : public Parent
{
public:
    Animal *getAnimal()
    {
        cout << "Child return dog instance" << endl;
        return new Dog();
    }
};
class Client
{
private:
    Parent *p;

public:
    Client(Parent *p) : p(p) {}

    void takeAnimal()
    {
        p->getAnimal();
    }
};
int main()
{
    Parent *parent = new Parent();
    Child *child = new Child();
    Client *client = new Client(child);
    client->takeAnimal();

    return 0;
}

File: Coder Army/SOLID/LSP(LiskovSubstitutionPrinciple)/Javascript code/LSP-Rules/SignatureRule/ReturnTypeRule.js
class Animal {
  // Common animal logic
}

class Dog extends Animal {
  // Dog-specific logic
}

class Parent {
  getAnimal() {
    console.log("Parent : Returning Animal instance");
    return new Animal();
  }
}

class Child extends Parent {
  getAnimal() {
    console.log("Child : Returning Dog instance");
    return new Dog();
  }
}

class Client {
  constructor(p) {
    this.p = p;
  }

  takeAnimal() {
    this.p.getAnimal();
  }
}

// Create Parent and Child instances
const parent = new Parent();
const child = new Child();

// Use Child
const client = new Client(child);
// Or use Parent
// const client = new Client(parent);

client.takeAnimal();
